src/Blocks/BlockPiston.cpp

```diff
@@ -86,105 +72,187 @@ bool cBlockPistonHandler::GetPlacementBlockTypeMeta(
 
 
 
-int cBlockPistonHandler::FirstPassthroughBlock(int a_PistonX, int a_PistonY, int a_PistonZ, NIBBLETYPE pistonmeta, cWorld * a_World)
+Vector3i cBlockPistonHandler::MetadataToOffset(NIBBLETYPE a_PistonMeta)
 {
-	// Examine each of the 12 blocks ahead of the piston:
-	for (int ret = 0; ret < PISTON_MAX_PUSH_DISTANCE; ret++)
+	switch (a_PistonMeta & 0x07)
 	{
-		BLOCKTYPE currBlock;
-		NIBBLETYPE currMeta;
-		AddPistonDir(a_PistonX, a_PistonY, a_PistonZ, pistonmeta, 1);
-		a_World->GetBlockTypeMeta(a_PistonX, a_PistonY, a_PistonZ, currBlock, currMeta);
-		if (cBlockInfo::IsPistonBreakable(currBlock))
+		case 0: return Vector3i( 0, -1,  0);
+		case 1: return Vector3i( 0,  1,  0);
+		case 2: return Vector3i( 0,  0, -1);
+		case 3: return Vector3i( 0,  0,  1);
+		case 4: return Vector3i(-1,  0,  0);
+		case 5: return Vector3i( 1,  0,  0);
+		default:
 		{
-			// This block breaks when pushed, extend up to here
-			return ret;
+			LOGWARNING("%s: invalid direction %d, ignoring", __FUNCTION__, a_PistonMeta & 0x07);
+			ASSERT(!"Invalid direction");
+			return Vector3i();
 		}
-		if (!CanPush(currBlock, currMeta))
+	}
+}
+
+
+
+
+
+void cBlockPistonHandler::PushBlocks(
+	const Vector3iSet & a_BlocksToPush,
+	cWorld * a_World, const Vector3i & a_PushDir
+)
+{
+	// Sort blocks to move the blocks first, which are farest away from the piston
+	// This prevents the overwriting of existing blocks
+	std::vector<Vector3i> sortedBlocks(a_BlocksToPush.begin(), a_BlocksToPush.end());
+	std::sort(sortedBlocks.begin(), sortedBlocks.end(), [a_PushDir](const Vector3i & a, const Vector3i & b)
+	{
+		return a.Dot(a_PushDir) > b.Dot(a_PushDir);
+	});
```

Comment:
Thats why I wasn't sure. I'm going to merge as is, and if we find its a perfomance issue, we can fix it later.